Client build work (HMR, build upgrades, refactoring asset placement)

dannon opened the pull request:
Fixes HMR build to be more reliable, and to work with styles -- you can now edit base.scss (or whatever) and it'll automatically reload an open browser with the new style applied.  

While getting this working I also upgraded a lot of the core devDependencies involved in the process, tweaked our webpack config a bit, and did some general refactoring in terms of build & static artifact placement trying to work towards a single standard 'dist' with less exposed.  I eliminated `june_2007_style` and adjusted the few things touching this.

Style sourcemap resolution should be fixed.

The general production build is marginally faster now.  (but you're really going to want to use the HMR build for developing now, anyway -- it's very fast comparatively).

Addressed comments on a standard for vue v-bind and v-on by enabling vue/strongly-recommended (with prettier compromises asserted) and standardizing the codebase on the recommended practice.

dannon committed:
Webpack version upgrades prior to extra build work.

dannon committed:
Babel upgrades

dannon committed:
Upgrade style-loader to 1.x

dannon committed:
Condense style loading rules

dannon committed:
Upgrade style building, adjust build args.

dannon committed:
Upgrade popper; deduplicate versions.

dannon committed:
Drop duplicate jquery alias.

dannon committed:
Deduplicate popper inclusions; minor ver update.

dannon committed:
Prettier (?) for webpack config.

dannon committed:
Upgrade vue related dependencies (exc: bootstrap-vue; will update that with bootstrap)

dannon committed:
Upgrade and rerun prettier.

dannon committed:
Upgrade eslint

dannon committed:
Fix eslint (drop unused var) in jobs.vue

dannon committed:
Update gulp and used dependencies.

dannon committed:
Upgrade miniCssExtract (trying to get HMR working...)

dannon committed:
Upgrade various loaders, major verions that drop some older node versions but they are fine with our pinned config.

dannon committed:
Minicssextract configuration changes.

dannon committed:
Build artifact refactoring to unify destinations to the standard 'dist' and allow old cruft to keep working.

dannon committed:
Update two (old, deprecated...) symlinks to base.css

dannon committed:
Drop old references to external JS -- all bundled now.

dannon committed:
Simplify js loading -- yagni

dannon committed:
Eliminate 'june_2007_style', squash some paths while working towards a flatter 'dist' exposure.  Rebuild config (includes mulled stuff that I guess we forgot to build before?)

nsoranzo commented on lib/galaxy/web/framework/helpers/__init__.py:
```diff
@@ -71,6 +71,15 @@ def css(*args):
 
+def dist_css(*args):
+    """
+    Transition function 'js' helper -- this is the modern way where all bundled
```
```suggestion
    Transition function 'css' helper -- this is the modern way where all bundled
```

nsoranzo commented on templates/webapps/tool_shed/repository/common.mako:
```diff
@@ -1002,7 +1002,7 @@
     %else:
         <td>
-            <img src="${h.url_for('/static')}/june_2007_style/blue/ok_small.png"/>
+            <img src="${h.url_for('/static')}/ok_small.png"/>
```
```suggestion
            <img src="${h.url_for('/static')}/style/ok_small.png"/>
```

nsoranzo commented on doc/source/admin/galaxy_options.rst:
```diff
@@ -1025,6 +1025,40 @@
+:Description:
+    Data directory used by beaker for caching mulled resolution
+    requests.
+:Default: ``citations/data``
```
See my comments on https://github.com/galaxyproject/galaxy/pull/8741 from today.

dannon replied:
I'll rebase over that when it's fixed in dev.  I don't know what the defaults should be or I'd do it now -- just `mulled/data` `mulled/locks` ?

nsoranzo replied:
Neither do I! It was just a heads up.

dannon committed:
Client readme tweaks to address comments from @nsoranzo in
https://github.com/galaxyproject/galaxy/pull/9122

dannon committed:
Apply suggestions from code review

Co-Authored-By: Nicola Soranzo <nicola.soranzo@gmail.com>

dannon committed:
Upgrade testing devDependencies.

dannon committed:
Transpiling updates prior to fixing tests (again..)

dannon committed:
Add modern qunit

dannon committed:
Drop old qunitjs

dannon committed:
Test a big sinon upgrade.

dannon committed:
Remove page qunit tests (component works, is well tested by selenium and more appropriate integration tests -- the 'page' is a massive 'unit' and these aren't easily tested in their current incarnation.  Good news is the layout is being rewritten right now.

dannon committed:
Wide swath of mostly minor updates

dannon committed:
Upgrade mocha to current major (+2)

dannon committed:
Cleanup.

dannon committed:
Alias to deduplicate underscore versions in bundle.

dannon committed:
Fix toolshed vue component test import.

dannon committed:
Attempt to branch out eslint vue styling rules, spawned from recent discussion on style choices in PR reviews.

dannon committed:
Add shorthand style for v-on

dannon committed:
Standardize v-bind syntax.

dannon committed:
Convert template attributes to kebab case per https://vuejs.org/v2/guide/components-props.html\#Prop-Casing-camelCase-vs-kebab-case.  Enforce via lint.

dannon committed:
Fix index variable shadowing in jobparametesrs, resolves final vue/no-template-shadow

dannon committed:
Update eslintrc.js, swap from 'vue/essential' to 'vue/strongly-recommended', enabling more rules.

dannon committed:
CamelCase js vars

dannon committed:
Merge remote-tracking branch 'upstream/dev' into client-build-work

dannon committed:
Fix attribute-hyphenation lint issues in merge.

dannon committed:
Fix variable shadowing of 'workflow' in new workflow toolbox.  Is the parent assigned 'workflow' ever used?  Looks like we can maybe just drop it, but this is safer for right now.

dannon committed:
Sync formatting post-merge.

dannon committed:
We don't use prettier via eslint; drop these unused dependencies.

dannon committed:
Swap several client dependencies to devDependencies.

dannon committed:
Swap galaxy.panels.mako to rely on new dist-css location.

dannon committed:
Mocha bump.

dannon committed:
Use flushpromises to fix test with back-to-back nextTick() tries;
standardize npm/yarn invocation.  todo: use the var that automatically
selects the invoking executable.

dannon committed:
Fix ToolSection test; add ticks between actions.

dannon committed:
Fix RepositoryTools tests, adding nextTicks after click events.

dannon committed:
Fix Repositories searchlist test

dannon committed:
Consistently specify yarn in instead of mentioning npm.  Use of npm bypasses the yarn lock and different versions of dependencies can cause issues.

dannon committed:
Add back test for Repositories list checking for no repos

dannon committed:
Merge remote-tracking branch 'upstream/dev' into client-build-work

dannon committed:
Standardize imports (shorthand dropping .js)

dannon commented:
Alright, this should work now, tests have been fixed.

dannon committed:
Merge remote-tracking branch 'upstream/dev' into client-build-work

dannon committed:
Fix/standardize v-bind-style and attribute-hyphenation in new upload components from merge.

dannon committed:
Fix component/local_iter shadowing in new upload components.

dannon committed:
Rerun prettier w/ new version.

dannon committed:
Drop -- for forwarding options via yarn (warning said it'll be passed literally in the future, but it does nothing post 1.0)

dannon committed:
Add ticks after initial mount prior to first click for new upload tests.

dannon committed:
Fix backbone import case

dannon committed:
Merge remote-tracking branch 'upstream/dev' into client-build-work

bgruening commented:
@OlegZharkov can you look at this and try it, please.

OlegZharkov commented:
> @OlegZharkov can you look at this and try it, please.

Sure. Do you also want me to merge it with https://github.com/galaxyproject/galaxy/pull/9261?

bgruening commented:
@OlegZharkov no. Just test this. It should be superuseful for you and @dannon appreciates some testing.

OlegZharkov commented:
@bgruening sure, np.
